python: propagate makeWrapper arguments

rnhmjoj opened the pull request:
###### Motivation for this change
Python libraries that have runtime dependencies, like Qt and GTK
libraries, are currently not working. These are not included in
the interpreter wrapper and no mechanism exist to do so automatically.

This commit adds a `nix-support/make-wrapper-args` file that stores
the makeWrapper arguments of a python derivation (created with
buildPythonPackage). The arguments are loaded from python.buildEnv
when creating the Python environment and passed to the makeWrapper
invocation.


###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change
- [ ] Tested compilation of all pkgs that depend on this change: obviously not.
- [x] Tested the wrapping is working for `pythonPackages.matplotlib` with Qt
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This should fix issues #25351 #80147 #39637

rnhmjoj committed:
pythonPackages.matplotlib: add GTK and Qt wrapper hooks

FRidh commented on pkgs/development/interpreters/python/mk-python-derivation.nix:
```diff
@@ -157,6 +157,19 @@ let
   ] ++ checkInputs;
 
   postFixup = lib.optionalString (!dontWrapPythonPrograms) ''
+    # Propagate wrapper arguments up to the python
```
In `mk-python-derivation` a hook needs to do this; the function itself should remain small.

@worldofpeace @jtojnar any common approach that can be taken here? I want declarative wrappers in JSON with structured attributes, but we're not there. Let's discuss this in the main thread.

rnhmjoj replied:
If the various hook were to export their arguments in a common array, be it `makeWrapperArgs` or something else it would be great.

rnhmjoj replied:
Ok, I moved the shell code to a dedicated hook. I tested again with matplotlib and it seems good. I used a clone of `buildPythonPackage` to avoid a mass rebuild, so I haven't tested much else.

FRidh reviewed (commented):
I think this is a good idea but want the bash code that's now in the `postFixup` somewhere else. 

My idea is that the various wrapper scripts write to this `make-wrapper-args` file, until we have declarative wrappers. The consumer of that file should maybe be able to choose what parts to take based on an identifier, though I think we can do without. Let's wait for more feedback.

rnhmjoj committed:
python: propagate makeWrapper arguments

Python libraries that have runtime dependencies, like Qt and GTK
libraries, are currently not working. These are not included in
the interpreter wrapper and no mechanism exist to do so automatically.

This commit adds a `nix-support/make-wrapper-args` file that stores
the makeWrapper arguments of a python derivation (created with
buildPythonPackage). The arguments are loaded from python.buildEnv
when creating the Python environment and passed to the makeWrapper
invocation.

rnhmjoj commented:
I can't reproduce the ofBorg failure with the eval command [here](https://github.com/nixos/ofborg#how-does-ofborg-call-nix-instantiate). Do you what's going on?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/16

doronbehar commented on pkgs/top-level/python-packages.nix:
```diff
@@ -4321,6 +4322,7 @@ in {
     stdenv = if stdenv.isDarwin then pkgs.clangStdenv else pkgs.stdenv;
     inherit (pkgs.darwin.apple_sdk.frameworks) Cocoa;
     inherit (pkgs) pkgconfig;
+    inherit (pkgs.qt5) wrapQtAppsHook;
```
@rnhmjoj according to my tests, this line is what's causing ofborg's eval error and it causes this:

```
env NIX_PATH=nixpkgs=$PWD nix-shell -p 'pythonPackages.matplotlib.override {enableGtk3=true;}' --run 'python -c "import matplotlib.pyplot as plt; plt.plot([1,2,3],[4,5,6]); plt.show()"'
```

(The command that reproduces https://github.com/NixOS/nixpkgs/issues/25351 ) to fail in the same way.

rnhmjoj replied:
Oh, I see now: it's the python 2 version! I'll fix it by I guess it will be removed soon enough.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/declarative-wrappers-another-idea/6661/1

FRidh commented:
I saw you force-pushed undoing the change to a hook.

rnhmjoj commented:
@FRidh I must have messed up when committing from another device. I should have fixed it now, thank you.

veprbl commented on pkgs/development/python-modules/matplotlib/2.nix:
```diff
@@ -43,7 +43,9 @@ buildPythonPackage rec {
     ++ stdenv.lib.optional enableGhostscript ghostscript
-    ++ stdenv.lib.optional stdenv.isDarwin [ Cocoa ];
+    ++ stdenv.lib.optional stdenv.isDarwin Cocoa
+    ++ stdenv.lib.optional enableQt wrapQtAppsHook
```
Here you add to `buildInputs`, but for matptlotlib 3.x it goes to `nativeBuildInputs`

rnhmjoj replied:
Right, it should go to `nativeBuildInputs`.

veprbl replied:
Not resolved.

rnhmjoj replied:
Thank you, I messed up again when rebasing. I should be more careful when working from multiples computers.

rnhmjoj commented:
Rebased and fixed a conflict.

rnhmjoj commented:
@FRidh what can I do to help move this forward? I'd really like to have this before the next release.

doronbehar commented:
> what can I do to help move this forward?

My summer vacation has started, and I started writing an RFC for declarative wrappers, based on #85103 . My draft is at https://github.com/doronbehar/rfcs/blob/declarative-wrappers/rfcs/0075-declarative-wrappers.md

jonringer commented on pkgs/development/interpreters/python/hooks/propagate-wrapper-args.sh:
```diff
@@ -0,0 +1,33 @@
+    echo "Executing propagateWrapperArgsHook"
+
+    mkdir -p "$out/nix-support"
+    if [ ''${#qtWrapperArgs[@]} -ne 0 ]; then
```
is the `''$` escaping necessary within a .sh file?

rnhmjoj replied:
Nope, I must have forgot to remove it once I moved the code to an external hook. Thanks.

ttuegel commented on pkgs/development/interpreters/python/hooks/propagate-wrapper-args.sh:
```diff
@@ -0,0 +1,20 @@
+# python interpreter buildEnv environment.
+echo "Sourcing propagate-wrapper-args.sh"
+
+propagateWrapperArgsHook() {
```
That sounds like a good idea to me! I don't think it's a blocker for this pull request, right?

rnhmjoj commented:
Rebased, fixed a conflict and removed the unnecessary escaping.

FRidh commented:
I have some changes lined up regarding this, but most importantly now, this PR breaks `python3.tests`. Every interpreter has a `tests` attribute containing some environment tests.

FRidh commented on pkgs/development/interpreters/python/wrapper.nix:
```diff
@@ -22,29 +22,39 @@ let
             if [ -f "$prg" ]; then
               rm -f "$out/bin/$prg"
               if [ -x "$prg" ]; then
-                makeWrapper "$path/bin/$prg" "$out/bin/$prg" --set NIX_PYTHONPREFIX "$out" --set NIX_PYTHONEXECUTABLE ${pythonExecutable} --set NIX_PYTHONPATH ${pythonPath} ${if permitUserSite then "" else ''--set PYTHONNOUSERSITE "true"''} ${stdenv.lib.concatStringsSep " " makeWrapperArgs}
```
ah `--set NIX_PYTHONPREFIX "$out"` got lost

rnhmjoj replied:
Thank you, fixed.

rnhmjoj commented:
@FRidh can you push your changes here? I couldn't use `git am` to pull them in because the hash of the ancestor changed while rebasing.

FRidh commented:
My apologies for not following up with this. I understand the need for this feature, but I am not happy with the eventual approach. 
I am of the opinion the specific hooks should propagate a generic hook that does the final wrapping, and to get there that takes more effort. I started this off in https://github.com/NixOS/nixpkgs/pull/102949. Once that is in, it would be trivial to add this feature.

Declarative wrappers is just so much more work, that is definitely not something to wait for.

rnhmjoj commented:
No problem, I very well knew this approach wasn't the ultimate solution. I hoped it could be used as a stop gap to quickly fix python packages with Qt dependencies, but it seems there isn't much of an interest: there are only a couple open issues in Nixpkgs. 

Thank you for starting the work on unifying the wrapper hooks. There was much talk about this situation but nothing material had been proposed.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/low-effort-and-high-impact-tasks/21095/41
